### PR TITLE
Dev docker multi user

### PR DESCRIPTION
**Pull request by jourdain**

### CODE REVIEW

**Comment by jourdain** [comment]
@gecage952 let me know how it goes... I just did a translation from what I had before as I'm not sure I can run it locally anyway.

**Comment by jourdain** [comment]
The docker will still serve on `/` but if you want to serve on `/app` you might be able to set `-e TRAME_URL_PREFIX=/app`.

**Comment by gecage952** [comment]
Sorry, just to update. I will get to testing this as soon as I can. It's just been a little crazy with the holiday and then a conference for the first part of this week.

**Comment by gecage952** [comment]
Ok, apologies for the delay. I tried out the docker image, and I'm getting an issue where the Trame startup animation is playing, but the app never loads. If I try to refresh, I get an error in the console along the lines of:
```
Traceback (most recent call last):
  File "/deploy/server/venv/lib/python3.10/site-packages/aiohttp/web_protocol.py", line 452, in _handle_request
    resp = await request_handler(request)
  File "/deploy/server/venv/lib/python3.10/site-packages/aiohttp/web_app.py", line 543, in _handle
    resp = await handler(request)
  File "/deploy/server/venv/lib/python3.10/site-packages/wslink/backends/aiohttp/launcher.py", line 97, in handle_post
    return await self._waitForReady(session, request)
  File "/deploy/server/venv/lib/python3.10/site-packages/wslink/backends/aiohttp/launcher.py", line 109, in _waitForReady
    if self.process_manager.isReady(session, count):
  File "/deploy/server/venv/lib/python3.10/site-packages/wslink/launcher.py", line 480, in isReady
    if not self.isRunning(id) and count < 60:
  File "/deploy/server/venv/lib/python3.10/site-packages/wslink/launcher.py", line 468, in isRunning
    return self.processes[id].poll() is None
KeyError: '8cbd5f5c-3ecf-11ef-bd1b-0242ac110002'

```

Do you know anything about this? All of the dependencies for this project are publicly available, so you should be able to build and run it yourself. When running without Docker, the app behaves as expected.

File: dockerfiles/setup/apps.yml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+trame: 
+  cmd:
+    - python
+    - -m
+    - template_app.app
```
Comment (gecage952) [inline dockerfiles/setup/apps.yml:5]: I believe this should be `launcher_app.app`

**Review by gecage952 — CHANGES_REQUESTED** [review]

**Comment by jourdain** [comment]
Yes please fix that. I used what was in the readme. I’m off computer till
August.

**Comment by gecage952** [comment]
Just an update for when you get back. Ok, so the app itself is running nicely. After testing with various nginx configs for our vms, I think we actually will need the `EP_PATH` parameter working. The url prefix flag doesn't seem to work, but the rest of the container is working after I fixed some issues with the app itself (Completely my fault haha). I can provide the nginx config that we're using just for clarity to help test.

**Comment by gecage952** [comment]
Also, a question has popped up while testing this. Is there a way to guarantee that a specific user/browser gets the same application or at least the same state every time they connect. At the moment it seems that all of the state is lost on refresh, which causes issues with authentication as we are redirecting to a third party login form. I see that you added the session id to the redirect endpoint that is set up here, but we actually can't configure wildcard redirect urls in our oauth provider to be able to properly redirect to these dynamically generated session ids.

**Comment by jourdain** [comment]
Yes, you just need to use the same sessionURL across request in the URL/cookie logic. Right now, we just use the URL path, but a cookie one could be added.

**Comment by jourdain** [comment]
Side note, with the latest push on dockerhub, the `TRAME_URL_PREFIX` env variable should properly work now.

**Comment by gecage952** [comment]
> Side note, with the latest push on dockerhub, the `TRAME_URL_PREFIX` env variable should properly work now.

Nice. Thanks for the heads up.

**Comment by gecage952** [comment]
So, the cookie method would be setting the cookie to the session id (let's say trame_launcher_session) in the trame app (probably main) and then adding something like this to the apache config: 
```
RewriteCond %{HTTP_COOKIE} trame_launcher_session=([^;]+)
RewriteRule ^/(.*)$ http://${session-to-port:%1}/$1 [L]
```
I'm not super familiar with apache, so do you know what the rule would look like?

**Comment by jourdain** [comment]
The cookie will be managed on the client side directly, no server change involved. We will need to update trame-client to add such support. Otherwise, right now, you can use the url.

**Comment by gecage952** [comment]
Ok gotcha.

**Comment by gecage952** [comment]
> The cookie will be managed on the client side directly, no server change involved. We will need to update trame-client to add such support. Otherwise, right now, you can use the url.

Do you know what URL to navigate to to get a specific session? I was trying `/api/{session_id}` but that didn't seem to route properly. For a few attempts on Friday, I thought that using `?sessionid={some_id}` worked, but as I'm trying it now, it seems to just be creating new sessions.

**Comment by jourdain** [comment]
If you are asking to get back to the trame app, it will be `?sessionURL={ws_url}` where ws_url should be encoded from `wss://proper-host/proxy?sessionId=${id}&path=ws`.

**Comment by jourdain** [comment]
If you are asking for API endpoint (your own route), it will indeed be `/api/{session_id}/{your api path on trame process}`

**Comment by gecage952** [comment]
Hm, for some reason, it's just showing the trame loading animation when i try the `sessionURL` method.
Basically, with my host being `localhost:8080`, and my session id being `d2b2998e-58d1-11ef-bcb7-0242ac110002 `, I'm trying to navigate to: `http://localhost:8080/?sessionURL=wss%3A%2F%2Flocalhost%3A8080%2Fproxy%3FsessionId%3Dd2b2998e-58d1-11ef-bcb7-0242ac110002%26path%3Dws`

**Comment by jourdain** [comment]
`http://` => `ws://` not `wss://`

**Comment by gecage952** [comment]
Whoops, thanks for that. When I go to the correct url (`http://localhost:8080/?sessionURL=ws%3A%2F%2Flocalhost%3A8080%2Fproxy%3FsessionId%3Dd2b2998e-58d1-11ef-bcb7-0242ac110002%26path%3Dws`) it seems to open the websocket connection, but it's giving me the loading screen again. In the web console, I see `Connection error {code: -32000, message: 'Authentication failed'}`. Have you seen this before?

**Comment by jourdain** [comment]
I see... So you need to add `&secret={something_secret}` to that url as well. Basically the app/process was started with `--authKey something_secret` to prevent anyone with the sessionURL to connect if he was not the creator of that process who know that `something_secret`.

**Comment by jourdain** [comment]
You can get that `something_secret` from the process of trame directly. It is part of its cli args.

**Comment by gecage952** [comment]
Ok. I'll work with that then.

**Comment by jourdain** [comment]
Sorry that I forgot about it...

**Comment by gecage952** [comment]
Ok. So we've got some things working, but a couple of issues you might have some insight into. For context, we are running the trame docker container at `localhost:8080` and we are running our own proxy at `localhost` (it's just nginx). Navigating to either of those URLs properly creates a new session of the trame app. When I want to return to a specific session, if I navigate directly to `localhost:8080?sessionURL=whatever`, it works properly. However, when I try to go through our proxy ie `localhost?sessionURL=whatever`, we're running into issues. After checking the apache logs in the Trame docker container, our proxy is correctly sending the request to apache and gets the index file, however when the js script tries to open a websocket connection, it's trying to connect to `localhost` instead of `localhost:8080`.

I recognize that this is probably tough to follow in text, so I took a screenshot of the networking (not sure how much it will help either though, sorry)
Screenshot from 2024-08-14 12-31-24

The logs show that the initial request is correctly of the form:
`localhost?sessionURL=ws%3A%2F%2Flocalhost%3A8080%2Fproxy%3FsessionId%3D41db9df6-5a4e-11ef-8b77-0242ac110002%26path%3Dws&secret=pbcGONjF1ursLAi8 `
And this is getting passed correctly to apache.


Here's the nginx config if it helps:
```
location / {
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "Upgrade";

        if ($arg_state) {
            proxy_pass http://127.0.0.1:8080/api/$arg_state$request_uri;
            break;
        }
        if ($cookie_trame_launcher_session) {
            proxy_pass [http://127.0.0.1:8080?sessionURL=ws%3A%2F%2F127.0.0.1%3A8080%2Fproxy%3FsessionId%3D$cookie_trame_launcher_session%26path%3Dws&secret=$cookie_trame_session_key](http://127.0.0.1:8080/?sessionURL=ws%3A%2F%2F127.0.0.1%3A8080%2Fproxy%3FsessionId%3D$cookie_trame_launcher_session%26path%3Dws&secret=$cookie_trame_session_key);
            break;
        }
        proxy_pass http://127.0.0.1:8080/$request_uri;
    }

```

**Comment by gecage952** [comment]
Sorry, I know this is kinda hard to follow over text. Let me know if you need more info.

**Comment by jourdain** [comment]
If you want we can have a quick debug meeting together but I can also try to write a couple of things below.

**Comment by jourdain** [comment]
So normally you should see a POST request on /paraview. The response of that request should give you both the secret and sessionURL. 

If that fails, the system will fallback assuming you are not running a launcher and try to connect to /ws. 

From what I see it seems like the launcher was not detected.

**Comment by SergeyYakubov** [comment]
I think the problem is that we see POST request to Paraview every time. And every time it creates a new session, even if we provide sessionURL.

**Comment by jourdain** [comment]
> I think the problem is that we see POST request to Paraview every time. And every time it creates a new session, even if we provide sessionURL.

That is indeed an issue and will need investigation on why you can not reconnect to an existing process.

**Comment by gecage952** [comment]
Yeah, I think a meeting would probably help. Pretty much anytime tomorrow afternoon works with just the same people as the last one.

**Comment by jourdain** [comment]
I think I'm just missing your email

**Comment by gecage952** [comment]
Let me respond to the earlier meeting invite.

**Comment by SergeyYakubov** [comment]
One thing I noticed:
after seemingly same requests to apache (one directly from browser, one via nginx proxy)

`127.0.0.1:8080/?sessionURL=ws%3A%2F%2F127.0.0.1%2Fproxy%3FsessionId%3D99aa4c26-5a85-11ef-b55e-0242ac110003%26path%3Dws&secret=ish5Tmh2ZLMNsiLU"`

I get different redirects to proxy:

this is via nginx:
`127.0.0.1/proxy?sessionid=9e11ce06-5a85-11ef-b55e-0242ac110003&path=ws`

and this is direct:
`127.0.0.1/proxy?sessionId=99aa4c26-5a85-11ef-b55e-0242ac110003&path=ws`

the first one has wrong value for session id i(only first part of it - 9e11ce06). Any ideas? Where exactly is sessionURL processed?

**Comment by jourdain** [comment]
No, those are 2 different sessions. Like you said, you always start a new process for some reason.

**Comment by SergeyYakubov** [comment]
I know that they are different, the question is why the two requests with the same sessionURL send to /proxy with different session ids

**Comment by jourdain** [comment]
The reason of that discrepancy is due to the fact that the proxy (nginx) use the cookie to do the routing rather than the URL.
